pytorch_lightning/utilities/cli.py

```diff
@@ -61,28 +63,38 @@ def __init__(self, *args: Any, parse_as_dict: bool = True, **kwargs: Any) -> Non
 
     def add_lightning_class_args(
         self,
-        lightning_class: Union[Type[Trainer], Type[LightningModule], Type[LightningDataModule], Type[Callback]],
+        lightning_class: Union[Callable[..., Union[Trainer, LightningModule, LightningDataModule, Callback]],
+                               Type[Trainer], Type[LightningModule], Type[LightningDataModule], Type[Callback]],
         nested_key: str,
         subclass_mode: bool = False
     ) -> List[str]:
         """
         Adds arguments from a lightning class to a nested key of the parser
 
         Args:
-            lightning_class: Any subclass of {Trainer, LightningModule, LightningDataModule, Callback}.
+            lightning_class: A callable or any subclass of {Trainer, LightningModule, LightningDataModule, Callback}.
             nested_key: Name of the nested namespace to store arguments.
             subclass_mode: Whether allow any subclass of the given class.
         """
-        assert issubclass(lightning_class, (Trainer, LightningModule, LightningDataModule, Callback))
-        if issubclass(lightning_class, Callback):
-            self.callback_keys.append(nested_key)
-        if subclass_mode:
-            return self.add_subclass_arguments(lightning_class, nested_key, required=True)
-        return self.add_class_arguments(
-            lightning_class,
-            nested_key,
-            fail_untyped=False,
-            instantiate=not issubclass(lightning_class, Trainer),
+        if callable(lightning_class) and not inspect.isclass(lightning_class):
+            lightning_class = class_from_function(lightning_class)
+
+        if inspect.isclass(lightning_class) and issubclass(
+            cast(type, lightning_class), (Trainer, LightningModule, LightningDataModule, Callback)
+        ):
+            if issubclass(cast(type, lightning_class), Callback):
```

Comment:
You only need to `cast(type, lightning_class)` once

Reply:
Done :smiley: